File: src/main/java/org/dedda/bratwurst/test/TestFunctionRunner.java
package org.dedda.bratwurst.test;

import org.dedda.bratwurst.lang.BWClass;
import org.dedda.bratwurst.lang.BWFunction;
import org.dedda.bratwurst.lang.Program;
import org.dedda.bratwurst.lang.scope.Scope;
import org.dedda.bratwurst.parse.Parser;

import java.io.File;
import java.util.Arrays;
import java.util.Optional;

/**
 * Created by dedda on 1/23/16.
 *
 * @author dedda
 */
public class TestFunctionRunner {

    private TestRunner runner;

    public void run(TestRunner runner, String filename, String functionName) {
        this.runner = runner;
        BWClass.unregisterAll();
        Parser parser = new Parser(new File(filename));
        Program program = parser.parse();
        Optional<BWFunction> fopt = Arrays.stream(program.getFunctions()).filter(f -> f.getName().equals(functionName)).findFirst();
        if (!fopt.isPresent()) {
            fail("Function " + functionName + "doesn't exist!");
        }
        BWFunction function = fopt.get();
        Scope scope = new Scope(program, runner, this);
        function.run(scope);

    }

    public void fail(String message) {
        System.out.println(message);
        runner.stop();
    }

}

File: src/main/java/org/dedda/bratwurst/test/TestFunctionRunner.kt
package org.dedda.bratwurst.test

import org.dedda.bratwurst.lang.classes.BWClass
import org.dedda.bratwurst.lang.scope.Scope
import org.dedda.bratwurst.parse.Parser
import java.io.File

/**
 * Created by dedda on 1/23/16.
 *
 * @author dedda
 */
class TestFunctionRunner(private val runner: TestFileRunner) {

    fun run(filename: String, functionName: String) {
        BWClass.unregisterAll()
        val parser = Parser(File(filename))
        val program = parser.parse()
        val functionOptional = program.functions.stream()
            .filter { f -> f.name == functionName }
            .findFirst()
        if (!functionOptional.isPresent) {
            fail("Function " + functionName + "doesn't exist!")
        }
        val function = functionOptional.get()
        val scope = Scope(program, runner, this)
        function.run(scope)
    }

    fun addAssertionResult(result: AssertionResult) {
        runner.incrementAssertions()
        when (result) {
            is AssertionSuccess -> print(".")
            is AssertionError -> fail(result.message)
        }
    }

    private fun fail(message: String) {
        println(message)
        runner.stop()
    }
}